src/methods/scgpt_fine_tuned/script.py

```diff
@@ -0,0 +1,463 @@
+  result_binned_key="X_binned",  # the key in input_train.layers to store the binned data
+)
+
+preprocessor(input_train, batch_key=None)
+preprocessor(input_test, batch_key=None)
+
+all_counts = (
+  input_train.layers["X_binned"].A
+  if issparse(input_train.layers["X_binned"])
+  else input_train.layers["X_binned"]
+)
+genes = input_train.var["feature_name"].tolist()
+
+celltypes_labels = input_train.obs["celltype_id"].tolist()  # make sure count from 0
+celltypes_labels = np.array(celltypes_labels)
+
+batch_ids = np.zeros(len(input_train.obs), dtype=int)
+num_batch_types = len(set(batch_ids))
+
+(
+  train_data,
+  valid_data,
+  train_celltype_labels,
+  valid_celltype_labels,
+  train_batch_labels,
+  valid_batch_labels,
+) = train_test_split(
+  all_counts, celltypes_labels, batch_ids, test_size=0.1, shuffle=True
+)
+
+from torchtext.vocab import Vocab
+from torchtext._torchtext import Vocab as VocabPybind
+if model_dir is None:
+  vocab = Vocab(VocabPybind(genes + special_tokens, None))  # bidirectional lookup [gene <-> int]
+vocab.set_default_index(vocab["<pad>"])
+gene_ids = np.array(vocab(genes), dtype=int)
+
+tokenized_train = scgpt.tokenizer.tokenize_and_pad_batch(
+  train_data,
+  gene_ids,
+  max_len=max_seq_len,
+  vocab=vocab,
+  pad_token=pad_token,
+  pad_value=pad_value,
+  append_cls=True,  # append <cls> token at the beginning
+  include_zero_gene=hyperparameters["include_zero_gene"],
+)
+tokenized_valid = scgpt.tokenizer.tokenize_and_pad_batch(
+  valid_data,
+  gene_ids,
+  max_len=max_seq_len,
+  vocab=vocab,
+  pad_token=pad_token,
+  pad_value=pad_value,
+  append_cls=True,
+  include_zero_gene=hyperparameters["include_zero_gene"],
+)
+logger.info(
+  f"train set number of samples: {tokenized_train['genes'].shape[0]}, "
+  f"\n\t feature length: {tokenized_train['genes'].shape[1]}"
+)
+logger.info(
+  f"valid set number of samples: {tokenized_valid['genes'].shape[0]}, "
+  f"\n\t feature length: {tokenized_valid['genes'].shape[1]}"
+)
+
+
+### Load the pre-trained scGPT model
+
+ntokens = len(vocab)  # size of vocabulary
+model = scgpt.model.TransformerModel(
+  ntokens,
+  embsize,
+  nhead,
+  d_hid,
+  nlayers,
+  nlayers_cls=n_layers_cls,
+  n_cls=num_types if training_settings["CLS"] else 1,
+  vocab=vocab,
+  dropout=hyperparameters["dropout"],
+  pad_token=pad_token,
+  pad_value=pad_value,
+  do_mvc=hyperparameters["MVC"],
+  do_dab=training_settings["DAB"],
+  use_batch_labels=False,
+  num_batch_labels=num_batch_types,
+  domain_spec_batchnorm=hyperparameters["DSBN"],
+  input_emb_style="continuous",
+  n_input_bins=hyperparameters["n_bins"],
+  cell_emb_style="cls",
+  mvc_decoder_style="inner product",
+  ecs_threshold=hyperparameters["ecs_threshold"],
+  explicit_zero_prob=explicit_zero_prob,
+  use_fast_transformer=hyperparameters["fast_transformer"],
+  fast_transformer_backend="flash", # "linear" or "flash"
+  pre_norm=hyperparameters["pre_norm"],
+)
+
+if model_file is not None:
```

Comment:
I don't think this `if` is needed, this file should always exist

Reply:
You're right, I'll remove it.